tests/monotouch-test/CoreFoundation/ProxyTest.cs

```diff
@@ -42,5 +44,182 @@ public void Fields ()
 				Dlfcn.dlclose (lib);
 			}
 		}
+#if !__WATCHOS__ && !MONOMAC
+		[Test]
+		public void TestPACParsingScript ()
+		{
+			// get the path for the pac file, try to parse it and ensure that 
+			// our cb was called
+			string pacPath = Path.Combine (NSBundle.MainBundle.BundlePath, "example.pac");
+			NSError error = null;
+			var script = File.ReadAllText (pacPath);
+			var targetUri = new Uri ("http://docs.xamarin.com");
+			var proxies = CFNetwork.ExecuteProxyAutoConfigurationScript (script, targetUri, out error);
+			Assert.IsNull (error, "Null error");
+			Assert.AreEqual (1, proxies.Length, "Length");
+			// assert the data of the proxy, although we are really testing the js used
+			Assert.AreEqual (8080, proxies [0].Port, "Port");
+		}
+
+		[Test]
+		public void TestPACParsingScriptRunLoop ()
+		{
+			// get the path for the pac file, try to parse it and ensure that 
+			// our cb was called
+			CFProxy [] proxies = null;
+			NSError error = null;
+			NSObject cbClient = null;
+
+			string pacPath = Path.Combine (NSBundle.MainBundle.BundlePath, "example.pac");
+
+			var script = File.ReadAllText (pacPath);
+			var targetUri = new Uri ("http://docs.xamarin.com");
+			var runLoop = CFRunLoop.Current;
+
+			// get source, execute run loop in a diff thread
+
+			CFNetwork.CFProxyAutoConfigurationResultCallback cb = delegate (NSObject c, CFProxy [] l, NSError e) {
+				proxies = l;
+				error = e;
+				cbClient = c;
+				runLoop.Stop ();
+			};
+			var source = CFNetwork.ExecuteProxyAutoConfigurationScript (script, targetUri, cb, new CFStreamClientContext ());
+			using (var mode = new NSString ("Xamarin.iOS.Proxy")) {
+				runLoop.AddSource (source, mode);
+				// block until cb is called
+				runLoop.RunInMode (mode, double.MaxValue, false);
+				runLoop.RemoveSource (source, mode);
+				Assert.IsNull (cbClient, "Null client");
+				Assert.IsNull (error, "Null error");
+				Assert.IsNotNull (proxies, "Not null proxies");
+				Assert.AreEqual (1, proxies.Length, "Length");
+				// assert the data of the proxy, although we are really testing the js used
+				Assert.AreEqual (8080, proxies [0].Port, "Port");
+			}
+		}
+
+		[Test]
+		public void TestPACParsingAsync ()
+		{
+			CFProxy [] proxies = null;
+			NSError error = null;
+			NSObject cbClient = null;
+			bool done = false;
+			string pacPath = Path.Combine (NSBundle.MainBundle.BundlePath, "example.pac");
+
+			var script = File.ReadAllText (pacPath);
+			var targetUri = new Uri ("http://docs.xamarin.com");
+			
+			Exception ex;
+			bool foundProxies;
+			// similar to the other tests, but we want to ensure that the async/await API works
+			TestRuntime.RunAsync (DateTime.Now.AddSeconds (30), async () => {
```

Comment:
How does this asynchronous (?) test code interact with the asynchronous source code/is there a pattern for writing async tests?

Reply:
This is the pattern you can use to test async code within the MonoTouch tests without blocking the UI thread and getting into a deadlock. (There are not many tests like this in the tests).

Reply:
To be more precise, the method will ensure that the cb is executed async without blocking the main thread.